### ISSUE TITLE
Error in /include/dba/dba_pdo.php

### ISSUE THREAD

**Issue by ilu33** [redmatrix/hubzilla#606]
We tried using Hubzilla with HHVM's repo authoritative mode and HHVM's error log runs full with lines like this one: 
"Warning: feof() expects parameter 1 to be resource, null given in /var/www/hubzilla/include/dba/dba_pdo.php on line 138
Warning: fread() expects parameter 1 to be resource, null given in /var/www/hubzilla/include/dba/dba_pdo.php on line 139"
It seems that php never encountered this issue while HHVM stumbles over it.

**Comment by zotlabs** [comment]
The warning should now be silenced. It's quite possible that hiphop may have issues with postgres binary data under PDO (photos in particular) and may require special handling. The PDO driver returns an open stream resource which one queries with stdio interfaces to obtain the actually binary data.  A null result should not normally be returned by these queries.

**Comment by ilu33** [comment]
I've updated the issue while you were answering. I'm way too slow :-)

**Comment by zotlabs** [comment]
You're probably on your own with this one. If you've got repoauthoritative you can't have eval anywhere and we need asn.1 for converting foreign public keys. The HTMLPurifier issue indicates that HTMLPURIFIER_PREFIX is defined somewhere and I certainly didn't define it. It can't be undefined or redefined so you may need to figure out who is defining it and why. I may have a solution for the tokens setting page. But you would still need to verify (and perhaps fix) what is happening with postgres binary data.

**Comment by ilu33** [comment]
No, the logfile overrun regarding line 138 and 139 is still there. This stops us from further testing because after a short while the server hangs. The problem is the use of feof in line 138, a fix is on the way.

**Comment by zotlabs** [comment]
The warning is "null given" and I checked if $str was null and skipped that code. The problem may be that there are many conflicting definitions of null. In 'C' it is (char *) 0, in PHP it is 'undefined' and in postgres it's also undefined but implemented as something else. So you may need to change that check to '(! $str)' which should cover the situation for any definition of null. It also matches an empty string, but in this function we should never get an empty string (unless it's yet another definition of null). You could also use (! is_resource($str)) but I can't test this and there are notes that it may not always be reliable.

**Comment by fsteinel** [comment]
The hhvm issue [problem with feof !!](https://github.com/facebook/hhvm/issues/4447) does mention the php feof function... 
``` diff
diff --git a/include/dba/dba_pdo.php b/include/dba/dba_pdo.php
index e235c46..31c6718 100755
--- a/include/dba/dba_pdo.php
+++ b/include/dba/dba_pdo.php
@@ -135,8 +135,11 @@ class dba_pdo extends dba_driver {
        function unescapebin($str) {
                if($this->driver_dbtype === 'pgsql') {
                        $x = '';
-                       while(! feof($str)) {
-                               $x .= fread($str,8192);
+                       while(! ( is_null($str) == true || feof($str) == true)) {
+                               $x .= fread($str);
+                               if(is_null($str) == true) break;
+                               $stream_meta_data = stream_get_meta_data($str);
+                               if($stream_meta_data['unread_bytes'] <= 0) break;
                        }
                        if(substr($x,0,2) === '\\x') {
                                $x = hex2bin(substr($x,2));
```

Now the warning is gone.
Selecting Photo no photos are shown, also all images (e.g. profile picture) are replaced by the default image.
The hhhvm error.log does not mention any warnings or errors.

As you mentioned stream, i switched to [stream_get_contents](https://secure.php.net/manual/en/function.stream-get-contents.php)

``` diff
diff --git a/include/dba/dba_pdo.php b/include/dba/dba_pdo.php
index e235c46..31c6718 100755
--- a/include/dba/dba_pdo.php
+++ b/include/dba/dba_pdo.php
@@ -135,8 +135,11 @@ class dba_pdo extends dba_driver {
        function unescapebin($str) {
                if($this->driver_dbtype === 'pgsql') {
                        $x = '';
-                       while(! feof($str)) {
-                               $x .= fread($str,8192);
+                       while(! ( is_null($str) == true || feof($str) == true)) {
+                               $x .= stream_get_contents($str);
+                               if(is_null($str) == true) break;
+                               $stream_meta_data = stream_get_meta_data($str);
+                               if($stream_meta_data['unread_bytes'] <= 0) break;
                        }       
                        if(substr($x,0,2) === '\\x') {
                                $x = hex2bin(substr($x,2));
```
Still no images.
